### DISCUSSION TITLE
React Native + Firebase Integration With Web3Auth

### DISCUSSION THREAD

**Discussion by Tguntenaar** [Web3Auth/Web3Auth#547] [Help]
Hi, I followed this tutorial: [React Native Web3Auth](https://web3auth.io/docs/integration-builder?lang=react-native&chain=eth&customAuthentication=no&whitelabel=no&mfaParams=no&usingEmailPasswordless=no&rnWorkflowMode=bare&evmFramework=web3). I did manage to login with and retrieve a private key etc. Now I would to [integrate web3auth with Firebase](https://web3auth.io/docs/guides/firebase). However this specific tutorial isn't for React Native but for Firebase for the web. In React Native `signInWithPopup(auth, googleProvider)` does not work. Do you guys have a working example of such integration?

Extra info
In the firebase [docs](https://firebase.google.com/docs/auth/web/google-signin#redirect-mode) it says that on mobile devices instead of signInWithPopup siginInWithRedirect is preferred.

**Comment by shahbaz17** [comment]
Hey @Tguntenaar 

Have you tried `siginInWithRedirect`? What does it return? 

This could be achieved by using `JWT` as the provider for login in React Native and passing `id_token` to web3auth.

```js
web3auth.login({
  loginProvider: LoginProvider.JWT,
  redirectUrl: resolvedRedirectUrl,
  extraLoginOptions: {
    id_token: "your JWT token",
  },
});
```

**Reply by Tguntenaar** [reply]
Hi @shahbaz17 

Thank for the quick reply! It was my intention to retrieve that token from firebase and then pass it to web3auth. Where I get stuck is getting that token.

I did try signInWithRedirect.
```js
[TypeError: (0, _$$_REQUIRE(_dependencyMap[0], "@firebase/auth").signInWithRedirect) is not a function. (In '(0, _$$_REQUIRE(_dependencyMap[0], "@firebase/auth").signInWithRedirect)(auth, provider)', '(0, _$$_REQUIRE(_dependencyMap[0], "@firebase/auth").signInWithRedirect)' is undefined)]
```
Although it states that signInWithRedirect is not a function, VSCode lets me navigate to it.

**Comment by Tguntenaar** [comment]
For people also looking into integrating firebase with the `@web3auth/react-native-sdk`.
```js

import auth from 'firebase/auth';
import Web3Auth, {
  LOGIN_PROVIDER,
  OPENLOGIN_NETWORK,
} from '@web3auth/react-native-sdk';
import * as WebBrowser from '@toruslabs/react-native-web-browser';
import {
  GoogleSignin,
  statusCodes,
} from '@react-native-google-signin/google-signin';

// Sign in with Google
await GoogleSignin.hasPlayServices();
const userInfo = await GoogleSignin.signIn();
const { idToken } = userInfo;

// Sign in with Firebase
const googleCredential = GoogleAuthProvider.credential(idToken);
signInWithCredential(auth, googleCredential);

 const web3auth = new Web3Auth(WebBrowser, {
    clientId: clientIdWeb3Auth,
    network: OPENLOGIN_NETWORK.TESTNET,
  });
const scheme = 'YOUR_SCHEME'; // Or your desired app redirection scheme
const resolvedRedirectUrl = `${scheme}://openlogin`;

const clientIdWeb3Auth = 'YOUR_CLIENT_ID';

// Sign in with web3auth
const state = await web3auth.login({
  loginProvider: LOGIN_PROVIDER.JWT,
  redirectUrl: resolvedRedirectUrl,
  extraLoginOptions: {
    id_token: idToken,
    verifierIdField: 'sub', // same as your JWT Verifier ID
  },
});
```

**Reply by shahbaz17** [reply]
Thanks for sharing the code @Tguntenaar

**Reply by ealeksandrov7** [reply]
hey guys. Just wondering how does this work without providing a loginConfig/domain, etc..? I am having trouble making it work. In my case I am not using firebase but the other part is the same. Can you please expand on this snippet some more? Thanks in advance!

**Reply by Tguntenaar** [reply]
@ealeksandrov7 depending on the whether you are using expo or bare react native you can follow [this documentation](https://www.npmjs.com/package/@react-native-google-signin/google-signin).

I'm building in bare react native and used this code to configure
```js
GoogleSignin.configure({
  webClientId:
    'yourownwebclientid.apps.googleusercontent.com',
  offlineAccess: true,
});
```

**Comment by huyta** [comment]
Hello @Tguntenaar,

I just try your codes, but it does not work at my side.

Did you config verifier anywhere?

Thanks,
Huy
